src/collect_as.jl

```diff
@@ -104,14 +129,16 @@ Tries to construct a value of type `t` from the iterator `iterator`. The type `t
 must either be concrete, or a `UnionAll` without constraints.
 """
 function collect_as(::Type{T}, iterator) where {T<:FixedSizeArray}
-    spec = fsa_spec_from_type(T)::SpecFSA
     size_class = Base.IteratorSize(iterator)
     if size_class == Base.IsInfinite()
         throw(ArgumentError("iterator is infinite, can't fit infinitely many elements into a `FixedSizeArray`"))
     end
-    dim_count_int = dimension_count_of(size_class)
-    check_count_value(dim_count_int)
-    dim_count = Val(dim_count_int)::Val
-    len_stat = length_status(size_class)
-    collect_as_fsa_checked(iterator, spec, dim_count, len_stat)::T
+    mem = parent_type_with_default(T)
+    fsv = collect_as_fsv(mem, iterator)
+    output_dimension_count = checked_dimension_count_of(T, size_class)
```

Comment:
Throw `DimensionMismatch` early:

```suggestion
    output_dimension_count = checked_dimension_count_of(T, size_class)
    fsv = collect_as_fsv(mem, iterator)
```